s3fs/core.py

```diff
@@ -642,7 +647,7 @@ def mkdir(self, path, acl="", **kwargs):
     def rmdir(self, path, **kwargs):
         """ Remove empty key or bucket """
         bucket, key = split_path(path)
-        if (key and self.info(path)['Size'] == 0) or not key:
+        if (key and self.du(path, total=True) == 0) or not key:
```

Comment:
I guess this isn't quite right: if the path happens to contain keys of zero size, we'll delete anyway. Conversely, I don't think this can delete a bucket. I am tempted to accept the PR anyway as is, but if you can think of a quick fix and test, that would be welcome.

Reply:
Added here checking if there are any sub directories. Since the du underneath calls ls everything should come from dirs cache

Reply:
I understand, but there could be legitimate keys under this directory that happen to have zero size, no?

Reply:
yes i agree please see the follow up commit